internal/store/verticalpodautoscaler.go

```diff
@@ -0,0 +1,256 @@
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+
+				for _, c := range a.Status.Recommendation.ContainerRecommendations {
+					ms = append(ms, vpaResourcesToMetrics(c.ContainerName, c.LowerBound)...)
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_verticalpodautoscaler_container_status_recommendation_upper_bound",
+			Type: metric.Gauge,
+			Help: "Maximum resources the container can use before the VerticalPodAutoscaler updater evicts it.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+				if a.Status.Recommendation == nil || a.Status.Recommendation.ContainerRecommendations == nil {
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+
+				for _, c := range a.Status.Recommendation.ContainerRecommendations {
+					ms = append(ms, vpaResourcesToMetrics(c.ContainerName, c.UpperBound)...)
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_verticalpodautoscaler_container_status_recommendation_target",
+			Type: metric.Gauge,
+			Help: "Target resources the VerticalPodAutoscaler recommends for the container.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+				if a.Status.Recommendation == nil || a.Status.Recommendation.ContainerRecommendations == nil {
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+				for _, c := range a.Status.Recommendation.ContainerRecommendations {
+					ms = append(ms, vpaResourcesToMetrics(c.ContainerName, c.Target)...)
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_verticalpodautoscaler_container_status_recommendation_uncapped_target",
+			Type: metric.Gauge,
+			Help: "Target resources the VerticalPodAutoscaler recommends for the container ignoring bounds.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+				if a.Status.Recommendation == nil || a.Status.Recommendation.ContainerRecommendations == nil {
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+				for _, c := range a.Status.Recommendation.ContainerRecommendations {
+					ms = append(ms, vpaResourcesToMetrics(c.ContainerName, c.UncappedTarget)...)
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+	}
+)
+
+func vpaResourcesToMetrics(containerName string, resources v1.ResourceList) []*metric.Metric {
+	ms := []*metric.Metric{}
+	for resourceName, val := range resources {
+		switch resourceName {
+		case v1.ResourceCPU:
+			ms = append(ms, &metric.Metric{
+				LabelValues: []string{containerName, sanitizeLabelName(string(resourceName)), string(constant.UnitCore)},
+				Value:       float64(val.MilliValue()) / 1000,
+			})
+		case v1.ResourceStorage:
```

Comment:
is it really necessary to whitelist each type here? can't we just normalize to base units in any case and put the resource name in the `unit` label?

Reply:
@brancz I'm a little conflicted about this - I'm actually tempted to remove the `Storage` cases (if I'm honest, I copied and pasted this from the `kube_pod_container_resource_requests`).

At the moment the VPA only operates on CPU and memory, though there's a discussion about adding ephemeral storage (https://github.com/kubernetes/autoscaler/issues/1751). Those are the only metrics exposed by the Metrics Server AFAIK, so it's unlikely we'd need to whitelist every type of resource.

I'm also not sure how we'd generically normalize to base units - I'd consider the base unit of CPU to be a `core`, and the `resource.Quantity` only provides an `int` Value for that 😢 

Personally I'm in favour of keeping this consistent with how `kube_pod_container_resource_requests` works, though I don't feel strongly about this

Reply:
That's fair, I was just thinking if we can find a default case similar to `kube_pod_container_resource_requests`. But I'm also happy to just add it when we're actually at a point where this is a reality.